File: parser/src/test/java/com/github/fburato/highwheelmodules/bytecodeparser/classpath/ClassLoaderClassPathRootTest.java
package com.github.fburato.highwheelmodules.bytecodeparser.classpath;

import static org.assertj.core.api.Assertions.assertThat;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import com.github.fburato.highwheelmodules.model.bytecode.ElementName;

public class ClassLoaderClassPathRootTest {

    private ClassLoaderClassPathRoot testee;

    @BeforeEach
    public void setup() {
        this.testee = new ClassLoaderClassPathRoot(Thread.currentThread().getContextClassLoader());
    }

    @Test
    public void shouldReturnNoClassNames() {

        assertThat(this.testee.classNames()).isEmpty();
    }

    @Test
    public void shouldReturnsBytesForClassesVisibleToParentLoader() throws Exception {
        assertThat(this.testee.getData(ElementName.fromClass(ClassLoaderClassPathRootTest.class))).isNotNull();
        assertThat(Test.class.getName()).isNotNull();
    }

    @Test
    public void testReturnsNullForClassesNotVisibleToParentLoader() throws Exception {
        assertThat(this.testee.getData(ElementName.fromString("FooFoo"))).isNull();
    }

    @Test
    public void testReturnsNullForResourcesNotVisibleToParentLoader() throws Exception {
        assertThat(this.testee.getResource("not defined")).isNull();
    }

    @Test
    public void testReturnsInputStreamForResourcesVisibleToParentLoader() throws Exception {
        assertThat(this.testee.getResource("aresource.txt")).isNotNull();
    }

}

File: parser/src/test/scala/com/github/fburato/highwheelmodules/bytecodeparser/classpath/ClassLoaderClassPathRootTest.scala
package com.github.fburato.highwheelmodules.bytecodeparser.classpath

import com.github.fburato.highwheelmodules.bytecodeparser.TryMatchers._
import com.github.fburato.highwheelmodules.model.bytecode.ElementName
import org.mockito.scalatest.MockitoSugar
import org.scalatest.OneInstancePerTest
import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec

import java.io.InputStream
import scala.util.Success

class ClassLoaderClassPathRootTest
    extends AnyWordSpec
    with Matchers
    with MockitoSugar
    with OneInstancePerTest {

  private val testee = new ClassLoaderClassPathRoot(this.getClass.getClassLoader)

  "classNames should return empty" in {
    testee.classNames shouldEqual Success(Seq())
  }

  "getData" should {
    "return not null for classes visible to loader" in {
      testee.getData(
        ElementName.fromClass(classOf[ClassLoaderClassPathRootTest])
      ) should beSuccessWith[Option[InputStream]](maybeStream => maybeStream should not be empty)
    }

    "return null for classes not visible to loader" in {
      testee.getData(ElementName.fromString("FooFoo")) shouldBe Success(None)
    }
  }

  "getResource" should {
    "return not null for resources visible to loader" in {
      testee.getResource("aresource.txt") should beSuccessWith[Option[InputStream]](maybeStream =>
        maybeStream should not be empty
      )
    }

    "return null for resource not visible to loader" in {
      testee.getResource("not defined") shouldBe Success(None)
    }
  }
}
